In [1]:
!pip install datasets
!pip install transformers
!pip install evaluate
import evaluate

from datasets import load_dataset
imdbDataset = load_dataset("imdb")
yelpDataset = load_dataset("yelp_review_full")

from transformers import pipeline
bert_pipelineTrun = pipeline("sentiment-analysis", model="nlptown/bert-base-multilingual-uncased-sentiment", truncation=True)
distilbert_pipelineTrun = pipeline("sentiment-analysis", model="distilbert-base-uncased-finetuned-sst-2-english", truncation=True)

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 480.6/480.6 kB 33.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 11.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 179.3/179.3 kB 14.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 15.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 22.0 MB/s eta 0:00:00
  Attempting uninstall: fsspec
    Found existing installation: fsspec 2024.10.0
    Uninstalling fsspec-2024.10.0:
      Successfully uninstalled fsspec-2024.10.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
gcsfs 2024.10.0 requires fsspec==2024.10.0, but you have fsspec 2024.9.0 which is incompatible.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.0/84.0 kB 9.6 MB/s eta 0:00:00


/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


README.md:   0%|          | 0.00/7.81k [00:00<?, ?B/s]

train-00000-of-00001.parquet:   0%|          | 0.00/21.0M [00:00<?, ?B/s]

test-00000-of-00001.parquet:   0%|          | 0.00/20.5M [00:00<?, ?B/s]

unsupervised-00000-of-00001.parquet:   0%|          | 0.00/42.0M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/25000 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/25000 [00:00<?, ? examples/s]

Generating unsupervised split:   0%|          | 0/50000 [00:00<?, ? examples/s]

README.md:   0%|          | 0.00/6.72k [00:00<?, ?B/s]

train-00000-of-00001.parquet:   0%|          | 0.00/299M [00:00<?, ?B/s]

test-00000-of-00001.parquet:   0%|          | 0.00/23.5M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/650000 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/50000 [00:00<?, ? examples/s]

config.json:   0%|          | 0.00/953 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/669M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/39.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/872k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

Device set to use cuda:0


config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

Device set to use cuda:0


In [2]:
imdbReviews = imdbDataset["test"]["text"]
imdbLabels = imdbDataset["test"]["label"]

yelpReviews = yelpDataset["test"]["text"]
yelpTrueLabels = yelpDataset["test"]["label"]

bertImdbResults = bert_pipelineTrun(imdbReviews)
distilbertImdbResults = distilbert_pipelineTrun(imdbReviews)

bertYelpResults = bert_pipelineTrun(yelpReviews)
distilbertYelpResults = distilbert_pipelineTrun(yelpReviews)

In [3]:
yelpLabelMapping = {"0": 0, "1": 0, "2": 1, "3": 1, "4": 1}
yelpTrueLabels = [yelpLabelMapping[str(label)] for label in yelpTrueLabels]

bertImdbLabels = [result["label"] for result in bertImdbResults]
distilbertImdbLabels = [result["label"] for result in distilbertImdbResults]

bertImdbLabelMapping = {"1 star": 0, "2 stars": 0, "3 stars": 1, "4 stars": 1, "5 stars": 1}
distilbertImdbLabelMapping = {"NEGATIVE": 0, "POSITIVE": 1}
bertImdbLabels = [bertImdbLabelMapping[label] for label in bertImdbLabels]
distilbertImdbLabels = [distilbertImdbLabelMapping[label] for label in distilbertImdbLabels]

bertYelpLabels = [result["label"] for result in bertYelpResults]
distilbertYelpLabels = [result["label"] for result in distilbertYelpResults]

bertYelpLabelMapping = {"1 star": 0, "2 stars": 0, "3 stars": 1, "4 stars": 1, "5 stars": 1}
distilbertYelpLabelMapping = {"NEGATIVE": 0, "POSITIVE": 1}
distilbertYelpLabels = [distilbertYelpLabelMapping[label] for label in distilbertYelpLabels]
bertYelpLabels = [bertYelpLabelMapping[label] for label in bertYelpLabels]

In [4]:
accuracy_metric = evaluate.load("accuracy")
f1_metric = evaluate.load("f1")

bertAccuracy = accuracy_metric.compute(predictions=bertImdbLabels, references=imdbLabels)
bertF1 = f1_metric.compute(predictions=bertImdbLabels, references=imdbLabels)

print("IMDB Reviews (Truncating):")
print("BERT Accuracy (Truncated):", bertAccuracy["accuracy"])
print("BERT F1 Score (Truncated):", bertF1["f1"])

distilbertAccuracy = accuracy_metric.compute(predictions=distilbertImdbLabels, references=imdbLabels)
distilbertF1 = f1_metric.compute(predictions=distilbertImdbLabels, references=imdbLabels)

print("DistilBERT Accuracy (Truncated):", distilbertAccuracy["accuracy"])
print("DistilBERT F1 Score (Truncated):", distilbertF1["f1"])

IMDB Reviews:
BERT Accuracy (Truncated): 0.8556
BERT F1 Score (Truncated): 0.8650063570413582
DistilBERT Accuracy (Truncated): 0.89072
DistilBERT F1 Score (Truncated): 0.8874794069192751


In [5]:
bertYelpAccuracy = accuracy_metric.compute(predictions=bertYelpLabels, references=yelpTrueLabels)
bertYelpF1 = f1_metric.compute(predictions=bertYelpLabels, references=yelpTrueLabels, average="weighted")

print("Yelp Reviews (Truncating):")
print("BERT Accuracy (Truncated):", bertYelpAccuracy["accuracy"])
print("BERT F1 Score (Truncated):", bertYelpF1["f1"])

distilbertYelpAccuracy = accuracy_metric.compute(predictions=distilbertYelpLabels, references=yelpTrueLabels)
distilbertYelpF1 = f1_metric.compute(predictions=distilbertYelpLabels, references=yelpTrueLabels, average="weighted")

print("DistilBERT Accuracy (Truncated):", distilbertYelpAccuracy["accuracy"])
print("DistilBERT F1 Score (Truncated):", distilbertYelpF1["f1"])

Yelp Reviews:
BERT Accuracy (Truncated): 0.86278
BERT F1 Score (Truncated): 0.8623747864713464
DistilBERT Accuracy (Truncated): 0.82458
DistilBERT F1 Score (Truncated): 0.8263340259137584


In [44]:
#Chunking Method:
bert_pipelineChun = pipeline("sentiment-analysis", model="nlptown/bert-base-multilingual-uncased-sentiment", truncation=False)
distilbert_pipelineChun = pipeline("sentiment-analysis", model="distilbert-base-uncased-finetuned-sst-2-english", truncation=False)

Device set to use cuda:0
Device set to use cuda:0


In [52]:
imdbChunkTexts = []
for review in imdbDataset['test']:
    text = review['text']
    words = text.split()
    maxTokens = 128
    chunks = [words[i:i + maxTokens] for i in range(0, len(words), maxTokens)]
    imdbChunkTexts.extend([' '.join(chunk) for chunk in chunks])

yelpChunkTexts = []
for review in yelpDataset['test']:
    text = review['text']
    words = text.split()
    maxTokens = 30
    chunks = [words[i:i + maxTokens] for i in range(0, len(words), maxTokens)]
    yelpChunkTexts.extend([' '.join(chunk) for chunk in chunks])

In [53]:
bertImdbResultsChun = bert_pipelineChun(imdbChunkTexts)
distilbertImdbResultsChun = distilbert_pipelineChun(imdbChunkTexts)

bertYelpResultsChun = bert_pipelineChun(yelpChunkTexts)
distilbertYelpResultsChun = distilbert_pipelineChun(yelpChunkTexts)

RuntimeError: The size of tensor a (610) must match the size of tensor b (512) at non-singleton dimension 1

In [33]:
imdbLabelsChun = imdbDataset["test"]["label"]
yelpTrueLabelsChun = yelpDataset["test"]["label"]

yelpLabelMappingChun = {"0": 0, "1": 0, "2": 1, "3": 1, "4": 1}
yelpTrueLabelsChun = [yelpLabelMappingChun[str(label)] for label in yelpTrueLabelsChun]

bertImdbLabelsChun = [result["label"] for result in bertImdbResultsChun]
distilbertImdbLabelsChun = [result["label"] for result in distilbertImdbResultsChun]

bertImdbLabelMappingChun = {"1 star": 0, "2 stars": 0, "3 stars": 1, "4 stars": 1, "5 stars": 1}
distilbertImdbLabelMappingChun = {"NEGATIVE": 0, "POSITIVE": 1}
bertImdbLabelsChun = [bertImdbLabelMappingChun[label] for label in bertImdbLabelsChun]
distilbertImdbLabelsChun = [distilbertImdbLabelMappingChun[label] for label in distilbertImdbLabelsChun]

bertYelpLabelsChun = [result["label"] for result in bertYelpResultsChun]
distilbertYelpLabelsChun = [result["label"] for result in distilbertYelpResultsChun]

bertYelpLabelMappingChun = {"1 star": 0, "2 stars": 0, "3 stars": 1, "4 stars": 1, "5 stars": 1}
distilbertYelpLabelMappingChun = {"NEGATIVE": 0, "POSITIVE": 1}
distilbertYelpLabelsChun = [distilbertYelpLabelMappingChun[label] for label in distilbertYelpLabelsChun]
bertYelpLabelsChun = [bertYelpLabelMappingChun[label] for label in bertYelpLabelsChun]

In [40]:
bertAccuracyChun = accuracy_metric.compute(predictions=bertImdbLabelsChun, references=imdbLabels)
bertF1Chun = f1_metric.compute(predictions=bertImdbLabelsChun, references=imdbLabels)

print("IMDB Reviews (Chunking):")
print("BERT Accuracy (Truncated):", bertAccuracyChun["accuracy"])
print("BERT F1 Score (Truncated):", bertF1Chun["f1"])

distilbertAccuracyChun = accuracy_metric.compute(predictions=distilbertImdbLabelsChun, references=imdbLabels)
distilbertF1Chun = f1_metric.compute(predictions=distilbertImdbLabelsChun, references=imdbLabels)

print("DistilBERT Accuracy (Truncated):", distilbertAccuracyChun["accuracy"])
print("DistilBERT F1 Score (Truncated):", distilbertF1Chun["f1"])

ValueError: Mismatch in the number of predictions (1) and references (25000)

In [ ]:
bertYelpAccuracyChun = accuracy_metric.compute(predictions=bertYelpLabelsChun, references=yelpTrueLabelsChun)
bertYelpF1Chun = f1_metric.compute(predictions=bertYelpLabelsChun, references=yelpTrueLabels, average="weighted")

print("Yelp Reviews (Chunking):")
print("BERT Accuracy (Truncated):", bertYelpAccuracyChun["accuracy"])
print("BERT F1 Score (Truncated):", bertYelpF1Chun["f1"])

distilbertYelpAccuracyChun = accuracy_metric.compute(predictions=distilbertYelpLabelsChun, references=yelpTrueLabelsChun)
distilbertYelpF1Chun = f1_metric.compute(predictions=distilbertYelpLabelsChun, references=yelpTrueLabels, average="weighted")

print("DistilBERT Accuracy (Truncated):", distilbertYelpAccuracyChun["accuracy"])
print("DistilBERT F1 Score (Truncated):", distilbertYelpF1Chun["f1"])

---
# Truncated Method:
Models:
- "nlptown/bert-base-multilingual-uncased-sentiment"
- "distilbert-base-uncased-finetuned-sst-2-english"

Datasets:
- IMDB Reviews
- Yelp Reviews

##Results

    IMDB Reviews:

    BERT Accuracy (Truncated): 0.8556
    BERT F1 Score (Truncated): 0.8650063570413582
    DistilBERT Accuracy (Truncated): 0.89072
    DistilBERT F1 Score (Truncated): 0.8874794069192751

‎

    Yelp Reviews:

    Yelp -> BERT Accuracy (Truncated): 0.86278
    Yelp -> BERT F1 Score (Truncated): 0.8623747864713464
    Yelp -> DistilBERT Accuracy (Truncated): 0.82458
    Yelp -> DistilBERT F1 Score (Truncated): 0.8263340259137584
---

# Chunking Method:

Models:
- "nlptown/bert-base-multilingual-uncased-sentiment"
- "distilbert-base-uncased-finetuned-sst-2-english"

Datasets:
- IMDB Reviews
- Yelp Reviews

---




